### PR TITLE
[JAV-579]support strArray

### PR DESCRIPTION
**Pull request by mt-monster**
@WillemJiang @wujimin @liubao68  pls review it

### CODE REVIEW

**Comment by WillemJiang** [comment]
@mt-monster  Please create the SCB JIRA for this PR.

File: common/common-rest/src/test/java/io/servicecomb/common/rest/codec/param/TestCsvQueryProcessor.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package io.servicecomb.common.rest.codec.param;
```
Comment (WillemJiang) [inline common/common-rest/src/test/java/io/servicecomb/common/rest/codec/param/TestCsvQueryProcessor.java]: Please add the license header.

File: swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package io.servicecomb.swagger.generator.core.processor.annotation;
```
Comment (WillemJiang) [inline swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java]: License header please.

File: common/common-rest/src/main/java/io/servicecomb/common/rest/codec/param/QueryProcessorCreator.java [hunk 3]
```diff
@@ -26,10 +29,36 @@
...
+    String collectionFormat = ((QueryParameter) parameter).getCollectionFormat();
+
+    if (collectionFormat == null) {
+      return new QueryProcessor(parameter.getName(), targetType);
+    }
+    if (collectionFormat.equals("csv")) {
+      return new CsvQueryProcessor(parameter.getName(), targetType);
+    }
+    if (collectionFormat.equals("multi")) {
+      return new MultiQueryProcessor(parameter.getName(), targetType);
```
Comment (wujimin) [inline common/common-rest/src/main/java/io/servicecomb/common/rest/codec/param/QueryProcessorCreator.java]: what's the relationship between MultiQueryProcessor and QueryProcessor
their code are the same.

File: swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import io.swagger.models.parameters.Parameter;
+import io.swagger.models.parameters.QueryParameter;
+
+public class ApiParamAnnotationProcessor implements ParameterAnnotationProcessor {
+  @Override
+  public void process(Object annotation, OperationGenerator operationGenerator, int paramIdx) {
+    List<Parameter> providerParameters = operationGenerator.getProviderParameters();
+    String collectionFormat = ((ApiParam) annotation).collectionFormat();
+    Parameter parameter = providerParameters.get(paramIdx);
+    if (parameter instanceof QueryParameter) {
```
Comment (wujimin) [inline swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java]: ApiParam annotation not only take affect to QueryParameter, and not only can change collectionFormat

File: swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/OperationGenerator.java [hunk 5]
```diff
@@ -266,6 +270,15 @@ protected void scanMethodParameters() {
         context.getDefaultParamProcessor().process(this, paramIdx);
       }
     }
+
+    processByApiParamAnnotation(apiParams);
```
Comment (wujimin) [inline swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/OperationGenerator.java]: please use normal mechanism, register processor to context

File: swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java (hunk 4 above)
Comment (wujimin) [reply swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java]: image
we should add a new priority level as i told you in the first day.

Comment (mt-monster) [reply swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java]: i already use apiMap , is it ok ?

Comment (mt-monster) [reply swagger/swagger-generator/generator-core/src/main/java/io/servicecomb/swagger/generator/core/processor/annotation/ApiParamAnnotationProcessor.java]: ApiParam annotation not only take affect to QueryParameter, and not only can change collectionFormat---- right .but this is ok for this issue, so the left foundation need further more codes to cover.

File: common/common-rest/src/main/java/io/servicecomb/common/rest/codec/param/QueryProcessorCreator.java [hunk 6]
```diff
@@ -26,10 +29,33 @@
...
+    ParamValueProcessorCreatorManager.INSTANCE.register(PARAMTYPE, this);
+  }
+
+  @Override
+  public ParamValueProcessor create(Parameter parameter, Type genericParamType) {
+    JavaType targetType = TypeFactory.defaultInstance().constructType(genericParamType);
+
+    String collectionFormat = ((QueryParameter) parameter).getCollectionFormat();
+
+    if (collectionFormat == null) {
```
Comment (wujimin) [inline common/common-rest/src/main/java/io/servicecomb/common/rest/codec/param/QueryProcessorCreator.java]: why not use simpler code?
```
    if ("csv".equals(collectionFormat)) {
       return new CsvQueryProcessor(parameter.getName(), targetType);
     }
 
     return new QueryProcessor(parameter.getName(), targetType);
```

File: demo/demo-schema/src/main/java/io/servicecomb/demo/CodeFirstRestTemplate.java [hunk 7]
```diff
@@ -75,9 +77,24 @@ public void testCodeFirst(RestTemplate template, String microserviceName, String
       testTraceIdOnContextContainsTraceId(template, cseUrlPrefix);
 
       testRawJson(template, cseUrlPrefix);
+      testStrArray(template, cseUrlPrefix);
     }
   }
 
+  private void testStrArray(RestTemplate template, String cseUrlPrefix) {
+    String[] instrArray = {"a1,b1,c1"};
+    ArrayList result = template.getForObject(cseUrlPrefix + "testGetStrArray?str=a1,b1,c1", ArrayList.class);
```
Comment (wujimin) [inline demo/demo-schema/src/main/java/io/servicecomb/demo/CodeFirstRestTemplate.java]: compile warning?

**Comment by wujimin** [comment]
1.logic has problems
2.ApiParam annotation not only take affect to QueryParameter, and not only can change collectionFormat
3.customers cancel csv requirement

so we put off this requirement

**Pull request closed without merging**